### ISSUE TITLE
memory allocation failure in TNEFFillMapi

### ISSUE THREAD

**Issue by asarubbo** [Yeraze/ytnef#42]
on 1.9.2:

```
# ytnefprint $FILE
==11998==ERROR: AddressSanitizer failed to allocate 0x2fffd03000 (206155296768) bytes of LargeMmapAllocator (error code: 12)
==11998==Process memory map follows:
        0x000000400000-0x00000053c000   /usr/bin/ytnefprint
        0x00000073c000-0x00000073d000   /usr/bin/ytnefprint
        0x00000073d000-0x000000742000   /usr/bin/ytnefprint
        0x000000742000-0x0000013a7000
        0x00007fff7000-0x00008fff7000
        0x00008fff7000-0x02008fff7000
        0x02008fff7000-0x10007fff8000
        0x600000000000-0x602000000000
        0x602000000000-0x602000010000
        0x602000010000-0x602e00000000
        0x602e00000000-0x602e00010000
        0x602e00010000-0x603000000000
        0x603000000000-0x603000010000
        0x603000010000-0x603e00000000
        0x603e00000000-0x603e00010000
        0x603e00010000-0x604000000000
        0x604000000000-0x604000010000
        0x604000010000-0x604e00000000
        0x604e00000000-0x604e00010000
        0x604e00010000-0x606000000000
        0x606000000000-0x606000010000
        0x606000010000-0x606e00000000
        0x606e00000000-0x606e00010000
        0x606e00010000-0x607000000000
        0x607000000000-0x607000010000
        0x607000010000-0x607e00000000
        0x607e00000000-0x607e00010000
        0x607e00010000-0x60b000000000
        0x60b000000000-0x60b000010000
        0x60b000010000-0x60be00000000
        0x60be00000000-0x60be00010000
        0x60be00010000-0x60f000000000
        0x60f000000000-0x60f000010000
        0x60f000010000-0x60fe00000000
        0x60fe00000000-0x60fe00010000
        0x60fe00010000-0x610000000000
        0x610000000000-0x610000010000
        0x610000010000-0x610e00000000
        0x610e00000000-0x610e00010000
        0x610e00010000-0x616000000000
        0x616000000000-0x616000010000
        0x616000010000-0x616e00000000
        0x616e00000000-0x616e00010000
        0x616e00010000-0x619000000000
        0x619000000000-0x619000010000
        0x619000010000-0x619e00000000
        0x619e00000000-0x619e00010000
        0x619e00010000-0x61d000000000
        0x61d000000000-0x61d000010000
        0x61d000010000-0x61de00000000
        0x61de00000000-0x61de00010000
        0x61de00010000-0x620000000000
        0x620000000000-0x620000010000
        0x620000010000-0x620e00000000
        0x620e00000000-0x620e00010000
        0x620e00010000-0x621000000000
        0x621000000000-0x621000010000
        0x621000010000-0x621e00000000
        0x621e00000000-0x621e00010000
        0x621e00010000-0x624000000000
        0x624000000000-0x624000010000
        0x624000010000-0x624e00000000
        0x624e00000000-0x624e00010000
        0x624e00010000-0x630000000000
        0x630000000000-0x630000020000
        0x630000020000-0x630e00000000
        0x630e00000000-0x630e00010000
        0x630e00010000-0x640000000000
        0x640000000000-0x640000003000
        0x7fe4582f7000-0x7fe458f00000
        0x7fe459000000-0x7fe459100000
        0x7fe459199000-0x7fe45b4eb000
        0x7fe45b4eb000-0x7fe45b67b000   /lib64/libc-2.23.so
        0x7fe45b67b000-0x7fe45b87a000   /lib64/libc-2.23.so
        0x7fe45b87a000-0x7fe45b87e000   /lib64/libc-2.23.so
        0x7fe45b87e000-0x7fe45b880000   /lib64/libc-2.23.so
        0x7fe45b880000-0x7fe45b884000
        0x7fe45b884000-0x7fe45b89a000   /usr/lib64/gcc/x86_64-pc-linux-gnu/6.3.0/libgcc_s.so.1
        0x7fe45b89a000-0x7fe45ba99000   /usr/lib64/gcc/x86_64-pc-linux-gnu/6.3.0/libgcc_s.so.1
        0x7fe45ba99000-0x7fe45ba9a000   /usr/lib64/gcc/x86_64-pc-linux-gnu/6.3.0/libgcc_s.so.1
        0x7fe45ba9a000-0x7fe45ba9b000   /usr/lib64/gcc/x86_64-pc-linux-gnu/6.3.0/libgcc_s.so.1
        0x7fe45ba9b000-0x7fe45ba9d000   /lib64/libdl-2.23.so
        0x7fe45ba9d000-0x7fe45bc9d000   /lib64/libdl-2.23.so
        0x7fe45bc9d000-0x7fe45bc9e000   /lib64/libdl-2.23.so
        0x7fe45bc9e000-0x7fe45bc9f000   /lib64/libdl-2.23.so
        0x7fe45bc9f000-0x7fe45bda3000   /lib64/libm-2.23.so
        0x7fe45bda3000-0x7fe45bfa2000   /lib64/libm-2.23.so
        0x7fe45bfa2000-0x7fe45bfa3000   /lib64/libm-2.23.so
        0x7fe45bfa3000-0x7fe45bfa4000   /lib64/libm-2.23.so
        0x7fe45bfa4000-0x7fe45bfaa000   /lib64/librt-2.23.so
        0x7fe45bfaa000-0x7fe45c1aa000   /lib64/librt-2.23.so
        0x7fe45c1aa000-0x7fe45c1ab000   /lib64/librt-2.23.so
        0x7fe45c1ab000-0x7fe45c1ac000   /lib64/librt-2.23.so
        0x7fe45c1ac000-0x7fe45c1c3000   /lib64/libpthread-2.23.so
        0x7fe45c1c3000-0x7fe45c3c2000   /lib64/libpthread-2.23.so
        0x7fe45c3c2000-0x7fe45c3c3000   /lib64/libpthread-2.23.so
        0x7fe45c3c3000-0x7fe45c3c4000   /lib64/libpthread-2.23.so
        0x7fe45c3c4000-0x7fe45c3c8000
        0x7fe45c3c8000-0x7fe45c400000   /usr/lib64/libytnef.so.0
        0x7fe45c400000-0x7fe45c5ff000   /usr/lib64/libytnef.so.0
        0x7fe45c5ff000-0x7fe45c600000   /usr/lib64/libytnef.so.0
        0x7fe45c600000-0x7fe45c620000   /usr/lib64/libytnef.so.0
        0x7fe45c620000-0x7fe45c643000   /lib64/ld-2.23.so
        0x7fe45c6c0000-0x7fe45c833000
        0x7fe45c833000-0x7fe45c843000
        0x7fe45c843000-0x7fe45c844000   /lib64/ld-2.23.so
        0x7fe45c844000-0x7fe45c845000   /lib64/ld-2.23.so
        0x7fe45c845000-0x7fe45c846000
        0x7ffcfd211000-0x7ffcfd232000   [stack]
        0x7ffcfd27e000-0x7ffcfd280000   [vvar]
        0x7ffcfd280000-0x7ffcfd282000   [vdso]
        0xffffffffff600000-0xffffffffff601000   [vsyscall]
==11998==End of process memory map.
==11998==AddressSanitizer CHECK failed: /tmp/portage/sys-libs/compiler-rt-sanitizers-4.0.0/work/compiler-rt-4.0.0.src/lib/sanitizer_common/sanitizer_common.cc:120 "((0 && "unable to mmap")) != (0)" (0x0, 0x0)
    #0 0x4d95cf in __asan::AsanCheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) /tmp/portage/sys-libs/compiler-rt-sanitizers-4.0.0/work/compiler-rt-4.0.0.src/lib/asan/asan_rtl.cc:69
    #1 0x4f4335 in __sanitizer::CheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) /tmp/portage/sys-libs/compiler-rt-sanitizers-4.0.0/work/compiler-rt-4.0.0.src/lib/sanitizer_common/sanitizer_termination.cc:79
    #2 0x4e3962 in __sanitizer::ReportMmapFailureAndDie(unsigned long, char const*, char const*, int, bool) /tmp/portage/sys-libs/compiler-rt-sanitizers-4.0.0/work/compiler-rt-4.0.0.src/lib/sanitizer_common/sanitizer_common.cc:120
    #3 0x4ed265 in __sanitizer::MmapOrDie(unsigned long, char const*, bool) /tmp/portage/sys-libs/compiler-rt-sanitizers-4.0.0/work/compiler-rt-4.0.0.src/lib/sanitizer_common/sanitizer_posix.cc:132
    #4 0x424c6a in __sanitizer::LargeMmapAllocator<__asan::AsanMapUnmapCallback>::Allocate(__sanitizer::AllocatorStats*, unsigned long, unsigned long) /tmp/portage/sys-libs/compiler-rt-sanitizers-4.0.0/work/compiler-rt-4.0.0.src/lib/asan/../sanitizer_common/sanitizer_allocator_secondary.h:41
    #5 0x424c6a in __sanitizer::CombinedAllocator<__sanitizer::SizeClassAllocator64<__asan::AP64>, __sanitizer::SizeClassAllocatorLocalCache<__sanitizer::SizeClassAllocator64<__asan::AP64> >, __sanitizer::LargeMmapAllocator<__asan::AsanMapUnmapCallback> >::Allocate(__sanitizer::SizeClassAllocatorLocalCache<__sanitizer::SizeClassAllocator64<__asan::AP64> >*, unsigned long, unsigned long, bool, bool) /tmp/portage/sys-libs/compiler-rt-sanitizers-4.0.0/work/compiler-rt-4.0.0.src/lib/asan/../sanitizer_common/sanitizer_allocator_combined.h:70
    #6 0x424c6a in __asan::Allocator::Allocate(unsigned long, unsigned long, __sanitizer::BufferedStackTrace*, __asan::AllocType, bool) /tmp/portage/sys-libs/compiler-rt-sanitizers-4.0.0/work/compiler-rt-4.0.0.src/lib/asan/asan_allocator.cc:407
    #7 0x41f1fb in __asan::Allocator::Calloc(unsigned long, unsigned long, __sanitizer::BufferedStackTrace*) /tmp/portage/sys-libs/compiler-rt-sanitizers-4.0.0/work/compiler-rt-4.0.0.src/lib/asan/asan_allocator.cc:605
    #8 0x41f1fb in __asan::asan_calloc(unsigned long, unsigned long, __sanitizer::BufferedStackTrace*) /tmp/portage/sys-libs/compiler-rt-sanitizers-4.0.0/work/compiler-rt-4.0.0.src/lib/asan/asan_allocator.cc:786
    #9 0x4cf7ba in calloc /tmp/portage/sys-libs/compiler-rt-sanitizers-4.0.0/work/compiler-rt-4.0.0.src/lib/asan/asan_malloc_linux.cc:75
    #10 0x7fe45c3e4e53 in TNEFFillMapi /tmp/ytnef-1.9.2/lib/ytnef.c:424:19
    #11 0x7fe45c3e1384 in TNEFMapiProperties /tmp/ytnef-1.9.2/lib/ytnef.c:396:7
    #12 0x7fe45c3f6b47 in TNEFParse /tmp/ytnef-1.9.2/lib/ytnef.c:1184:15
    #13 0x7fe45c3f59d3 in TNEFParseFile /tmp/ytnef-1.9.2/lib/ytnef.c:1042:10
    #14 0x508814 in main /tmp/ytnef-1.9.2/ytnefprint/main.c:80:9
    #15 0x7fe45b50b78f in __libc_start_main /tmp/portage/sys-libs/glibc-2.23-r3/work/glibc-2.23/csu/../csu/libc-start.c:289
    #16 0x419c38 in _start (/usr/bin/ytnefprint+0x419c38)
```

Testcase:
https://github.com/asarubbo/poc/blob/master/00246-ytnef-memallocfailures

**Comment by carnil** [comment]
This is CVE-2017-9473

**Comment by ohwgiles** [comment]
With this test case, `-65535` is passed as `p->count` in line `424`. There is already an `ALLOCCHECK` directly afterwards, so this won't lead to a segmentation fault, but the CVE correctly states this could be used in a memory consumption DoS attack. Solved by #57
